diff --git a/bdwaf/modsecurity/nginx/modsecurity/apr_bucket_nginx.c b/bdwaf/modsecurity/nginx/modsecurity/apr_bucket_nginx.c
index 62eb59a..1ee7585 100644
--- a/bdwaf/modsecurity/nginx/modsecurity/apr_bucket_nginx.c
+++ b/bdwaf/modsecurity/nginx/modsecurity/apr_bucket_nginx.c
@@ -208,13 +208,14 @@ move_brigade_to_chain(apr_bucket_brigade *bb, ngx_chain_t **ll, ngx_pool_t *pool
                 if (cl == NULL) {
                     break;
                 }
-                
+
                 cl->buf = ngx_calloc_buf(pool);
                 if (cl->buf == NULL) {
                     break;
                 }
 
                 cl->buf->last_buf = 1;
+                cl->next = NULL;//Added for post_response_500 in 2017-5-17
                 *ll = cl;
             } else {
                 cl->buf->last_buf = 1;
diff --git a/bdwaf/modsecurity/nginx/modsecurity/ngx_http_modsecurity.c b/bdwaf/modsecurity/nginx/modsecurity/ngx_http_modsecurity.c
index 3a192f3..494a265 100644
--- a/bdwaf/modsecurity/nginx/modsecurity/ngx_http_modsecurity.c
+++ b/bdwaf/modsecurity/nginx/modsecurity/ngx_http_modsecurity.c
@@ -440,7 +440,7 @@ ngx_http_modsecurity_save_headers_in_visitor(void *data, const char *key, const
     hh = ngx_hash_find(&cmcf->headers_in_hash, h->hash,
                        h->lowcase_key, h->key.len);
 
-    // add by wuyao721 start 
+    // add by wuyao721 start,2017-5-8
     if (strncmp((char *)h->lowcase_key, "host", h->key.len) == 0){
          if (r->headers_in.host == NULL) {
                   r->headers_in.host = h;
@@ -448,7 +448,7 @@ ngx_http_modsecurity_save_headers_in_visitor(void *data, const char *key, const
          r->headers_in.server = h->value;
         return 1;
      }
-    // add by wuyao721 end   
+    // add by wuyao721 end
 
     if (hh && hh->handler(r, h, hh->offset) != NGX_OK) {
         return 0;
@@ -744,11 +744,13 @@ static int
 ngx_http_modsecurity_save_headers_out_visitor(void *data, const char *key, const char *value)
 {
     ngx_http_request_t             *r = data;
-    ngx_table_elt_t                *h, he;
-    ngx_http_upstream_header_t     *hh;
-    ngx_http_upstream_main_conf_t  *umcf;
+    ngx_table_elt_t                *h, he, *new_h;//add *new_h 2017-5-17
+    /*comment start, for post_response_500 in 2017-5-17*/
+    //ngx_http_upstream_header_t     *hh;
+    //ngx_http_upstream_main_conf_t  *umcf;
 
-    umcf = ngx_http_get_module_main_conf(r, ngx_http_upstream_module);
+    //umcf = ngx_http_get_module_main_conf(r, ngx_http_upstream_module);
+    /*comment end, 2017-5-17*/
 
     h = &he;
 
@@ -765,18 +767,30 @@ ngx_http_modsecurity_save_headers_out_visitor(void *data, const char *key, const
 
     ngx_strlow(h->lowcase_key, h->key.data, h->key.len);
 
-    h->hash = ngx_hash_key(h->lowcase_key, h->key.len);
+    h->hash = ngx_hash_key(h->lowcase_key, h->key.len);
 
-    hh = ngx_hash_find(&umcf->headers_in_hash, h->hash,
-                       h->lowcase_key, h->key.len);
+    /*comment start, for post_response_500 in 2017-5-17*/
+    //hh = ngx_hash_find(&umcf->headers_in_hash, h->hash,
+    //                   h->lowcase_key, h->key.len);
 
-    if (hh) {
+    //if (hh) {
         /* copy all */
         //modified by vincent
-        if (hh->copy_handler(r, h, hh->conf) != NGX_OK) {
-            return 0;
-        }
+        //if (hh->copy_handler(r, h, hh->conf) != NGX_OK) {
+        //    return 0;
+        //}
+    //}
+    /*comment end, 2017-5-17*/
+
+    /*add start,for post_response_500 in 2017-5-17*/
+    new_h = ngx_list_push(&r->headers_out.headers);
+    if (new_h == NULL) {
+        return NGX_ERROR;
     }
+    new_h->hash = h->hash;
+    new_h->key = h->key;
+    new_h->value = h->value;
+    /*add end, 2017-5-17*/
 
     ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
                    "ModSecurity: save headers out: \"%V: %V\"",
@@ -933,13 +947,13 @@ ngx_http_modsecurity_init(ngx_conf_t *cf)
         return NGX_ERROR;
     }
     *h = ngx_http_modsecurity_handler;
-    if( 0 ){
+
     ngx_http_next_header_filter = ngx_http_top_header_filter;
     ngx_http_top_header_filter = ngx_http_modsecurity_header_filter;
 
     ngx_http_next_body_filter = ngx_http_top_body_filter;
     ngx_http_top_body_filter = ngx_http_modsecurity_body_filter;
-    }
+
     ngx_memzero(&ngx_http_modsecurity_upstream, sizeof(ngx_http_upstream_t));
     ngx_http_modsecurity_upstream.cacheable = 1;
     return NGX_OK;
@@ -977,7 +991,7 @@ ngx_http_modsecurity_handler(ngx_http_request_t *r)
 
     /* create / retrive request ctx */
     if (r->internal) {
-        
+
         ctx = ngx_http_get_module_pool_ctx(r, ngx_http_modsecurity);
 
         if (ctx) {
@@ -1019,7 +1033,7 @@ ngx_http_modsecurity_handler(ngx_http_request_t *r)
         return NGX_DECLINED;
     }
 
-    if (r->method == NGX_HTTP_POST 
+    if (r->method == NGX_HTTP_POST
             && modsecIsRequestBodyAccessEnabled(ctx->req) ) {
 
         /* read POST request body, should we process PUT? */
@@ -1030,7 +1044,7 @@ ngx_http_modsecurity_handler(ngx_http_request_t *r)
 
         return NGX_DONE;
     }
-    
+
     /* other method */
     return ngx_http_modsecurity_status(r, modsecProcessRequestBody(ctx->req));
 }
@@ -1079,7 +1093,7 @@ ngx_http_modsecurity_header_filter(ngx_http_request_t *r) {
     ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
 
     /* already processed, checking redirect action. */
-    if (ctx && ctx->complete 
+    if (ctx && ctx->complete
                && r->err_status >= NGX_HTTP_MOVED_PERMANENTLY
                && r->err_status < 308) {
 
@@ -1142,12 +1156,12 @@ ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
         if (size) {
             char *data = apr_pmemdup(bb->p, buf->pos, size);
             if (data == NULL) {
-                return ngx_http_filter_finalize_request(r, 
+                return ngx_http_filter_finalize_request(r,
                          &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
             }
             e = apr_bucket_pool_create(data , size, bb->p, bb->bucket_alloc);
             if (e == NULL) {
-                return ngx_http_filter_finalize_request(r, 
+                return ngx_http_filter_finalize_request(r,
                          &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
             }
             APR_BRIGADE_INSERT_TAIL(bb, e);
@@ -1158,7 +1172,7 @@ ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
             buf->last_buf = 0;
             e = apr_bucket_eos_create(bb->bucket_alloc);
             if (e == NULL) {
-                return ngx_http_filter_finalize_request(r, 
+                return ngx_http_filter_finalize_request(r,
                          &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
             }
             APR_BRIGADE_INSERT_TAIL(bb, e);
@@ -1179,7 +1193,7 @@ ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
     if (ngx_http_modsecurity_load_headers_in(r) != NGX_OK
             || ngx_http_modsecurity_load_headers_out(r) != NGX_OK) {
 
-        return ngx_http_filter_finalize_request(r, 
+        return ngx_http_filter_finalize_request(r,
                  &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
     }
 
@@ -1193,14 +1207,14 @@ ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
 
     rc = move_brigade_to_chain(ctx->brigade, &out, r->pool);
     if (rc == NGX_ERROR) {
-        return ngx_http_filter_finalize_request(r, 
+        return ngx_http_filter_finalize_request(r,
                  &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
     }
 
     if (ngx_http_modsecurity_save_headers_in(r) != NGX_OK
             ||ngx_http_modsecurity_save_headers_out(r) != NGX_OK) {
 
-        return ngx_http_filter_finalize_request(r, 
+        return ngx_http_filter_finalize_request(r,
                  &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
     }
 
@@ -1347,7 +1361,7 @@ ngx_http_modsecurity_cleanup(void *data)
     if (ctx->connection != NULL) {
         (void) modsecFinishConnection(ctx->connection);
     }
-    
+
 }
 
     static char *
